File: CSharp/GetRealGoodUtils.Tests/ArraysTests.cs
using System;
using NUnit.Framework;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;

namespace GetRealGood.Utils.Tests
{
    public class ArraysTests
    {
        [OneTimeSetUp]
        public void SetupAll()
        {
        }

        [OneTimeTearDown]
        public void TearDownAll()
        {
        }

        [SetUp]
        public void Setup()
        {
        }

        [TearDown]
        public void TearDown()
        {
        }

        class IsEmptyTests
        {

            [Test]
            public void IsEmpty_Common_Array_Boolean()
            {

                try
                {
                    var o = new String[] {"sample1", "sample2"};
                    Assert.AreEqual(Arrays.IsEmpty(o), false);

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }
            }

            [Test]
            public void IsEmpty_Common_Array_Boolean__1()
            {

                try
                {
                    var o = new String[0];
                    Assert.AreEqual(Arrays.IsEmpty(o), true);

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }
            }

            [Test]
            public void IsEmpty_Special_Array_Boolean()
            {

                try
                {
                    var o = new String[] {"", ""};
                    Assert.AreEqual(Arrays.IsEmpty(o), false);

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }
            }

            [Test]
            public void IsEmpty_Special_Array_Boolean__1()
            {

                try
                {
                    var o = new String[] {null, null};
                    Assert.AreEqual(Arrays.IsEmpty(o), false);

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }
            }

            [Test]
            public void IsEmpty_Nullable_Null_ArraysException()
            {

                try
                {
                    String[] o = null;
                    Assert.Throws(typeof(ArraysException), delegate { Arrays.IsEmpty(o); });

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }
            }

        }


        class IsNotEmptyTests
        {

            [Test]
            public void IsNotEmpty_Common_Array_Boolean()
            {

                try
                {
                    var o = new String[] {"sample1", "sample2"};
                    Assert.AreEqual(Arrays.IsNotEmpty(o), true);

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }
            }

            [Test]
            public void IsNotEmpty_Common_Array_Boolean__1()
            {

                try
                {
                    var o = new String[0];
                    Assert.AreEqual(Arrays.IsNotEmpty(o), false);

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }
            }

            [Test]
            public void IsNotEmpty_Special_Array_Boolean()
            {

                try
                {
                    var o = new String[] {"", ""};
                    Assert.AreEqual(Arrays.IsNotEmpty(o), true);

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }
            }

            [Test]
            public void IsNotEmpty_Special_Array_Boolean__1()
            {

                try
                {
                    var o = new String[] {null, null};
                    Assert.AreEqual(Arrays.IsNotEmpty(o), true);

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }
            }

            [Test]
            public void IsNotEmpty_Nullable_Null_ArraysException()
            {

                try
                {
                    String[] o = null;
                    Assert.Throws(typeof(ArraysException), delegate { Arrays.IsNotEmpty(o); });

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }
            }

        }


        class HasNullsAtTests
        {

            [Test]
            public void HasNullsAt_Common_Array_List()
            {
                try
                {

                    String[] s = new String[] {null, null, "null"};

                    var result = Arrays.HasNullsAt(s);
                    Assert.AreEqual(2, result.Count());

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }

            }

            [Test]
            public void HasNullsAt_Common_Collection_List__1()
            {
                try
                {

                    String[] s = new String[] {"", "null", "null"};

                    var result = Arrays.HasNullsAt(s);
                    Assert.AreEqual(0, result.Count());

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }

            }

            [Test]
            public void HasNullsAt_Nullable_Null_ArraysException()
            {
                try
                {

                    String[] s = null;

                    Assert.Throws(typeof(ArraysException), delegate { Arrays.HasNullsAt(s); });

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }

            }

            [Test]
            public void HasNullsAt_Special_Array_List()
            {
                try
                {
                    String[] s = new String[] {null, null, null};

                    var result = Arrays.HasNullsAt(s);
                    Assert.AreEqual(3, result.Count());

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }

            }

        }


        class HasNullsTests
        {

            [Test]
            public void HasNulls_Common_Array_Boolean()
            {
                try
                {

                    String[] s = new String[] {"null", null, null};

                    var result = Arrays.HasNulls(s);
                    Assert.AreEqual(true, result);

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }
            }

            [Test]
            public void HasNulls_Common_Array_Boolean__1()
            {
                try
                {

                    String[] s = new String[] {"null", "null", "null"};

                    var result = Arrays.HasNulls(s);
                    Assert.AreEqual(false, result);

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }
            }

            [Test]
            public void HasNulls_Special_Array_Boolean()
            {
                try
                {
                    String[] s = new String[] {null, null, null};

                    var result = Arrays.HasNulls(s);
                    Assert.AreEqual(true, result);

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }
            }

            [Test]
            public void HasNulls_Nullable_Null_ArraysException()
            {
                try
                {

                    String[] s = null;

                    Assert.Throws(typeof(ArraysException), delegate { Arrays.HasNulls(s); });

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }
            }

        }


        class HasNullsCountTests
        {

            [Test]
            public void HasNullsCount_Common_Array_Long()
            {
                try
                {

                    String[] s = new String[] {"null", null, null};

                    var result = Arrays.HasNullsCount(s);
                    Assert.AreEqual(2, result);

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }

            }

            [Test]
            public void HasNullsCount_Common_Array_Long__1()
            {
                try
                {

                    String[] s = new String[] {"null", "null", "null"};

                    var result = Arrays.HasNullsCount(s);
                    Assert.AreEqual(0, result);

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }

            }

            [Test]
            public void HasNullsCount_Nullable_Null_ArraysException()
            {
                try
                {

                    String[] s = null;

                    Assert.Throws(typeof(ArraysException), delegate { Arrays.HasNullsCount(s); });

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }

            }

            [Test]
            public void HasNullsCount_Special_Array_Long()
            {
                try
                {

                    String[] s = new String[] {null, null, null};

                    var result = Arrays.HasNullsCount(s);
                    Assert.AreEqual(3, result);

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }

            }


        }


        class MaxTests
        {

            [Test]
            public void Max_Common_Collection_Object()
            {

                try
                {

                    var coll = new String[] {"a", "b", "c"};

                    var result = Arrays.Max(coll);

                    Assert.IsTrue(result.Equals("c"));

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }

            }



            [Test]
            public void Max_Nullable_Null_ArraysException()
            {

                try
                {

                    String[] coll = null;

                    Assert.Throws(typeof(ArraysException), delegate { Arrays.Max(coll); });

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }

            }

            [Test]
            public void Max_Special_Arrays_ArraysException()
            {

                try
                {

                    var coll = new String[] {null, null, null};

                    Assert.Throws(typeof(ArraysException), delegate { Arrays.Max(coll); });

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }

            }

            [Test]
            public void Max_Special_Array_ArraysException__1()
            {

                try
                {

                    var coll = new String[0];

                    Assert.Throws(typeof(ArraysException), delegate { Arrays.Max(coll); });

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }

            }

            [Test]
            public void Max_Special_Array_Object__2()
            {

                try
                {

                    var coll = new String[] {"a", "b", null};

                    var result = Arrays.Max(coll);

                    Assert.IsTrue(result.Equals("b"));

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }

            }

        }

        class MinTests
        {

            [Test]
            public void Min_Common_Array_Object()
            {

                try
                {

                    var coll = new String[] {"a", "b", "c"};

                    var result = Arrays.Min(coll);

                    Assert.IsTrue(result.Equals("a"));

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }

            }



            [Test]
            public void Min_Nullable_Null_ArraysException()
            {

                try
                {

                    String[] coll = null;

                    Assert.Throws(typeof(ArraysException), delegate { Arrays.Min(coll); });

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }

            }

            [Test]
            public void Min_Special_Array_ArraysException()
            {

                try
                {

                    var coll = new String[] {null, null, null};

                    Assert.Throws(typeof(ArraysException), delegate { Arrays.Min(coll); });

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }

            }

            [Test]
            public void Min_Special_Array_ArraysException__1()
            {

                try
                {

                    var coll = new String[0];

                    Assert.Throws(typeof(ArraysException), delegate { Arrays.Min(coll); });

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }

            }

            [Test]
            public void Min_Special_Array_Object__2()
            {

                try
                {

                    var coll = new String[] {"a", "b", null};

                    var result = Arrays.Min(coll);

                    Assert.IsTrue(result.Equals("a"));

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }

            }

        }

        class PeaksTests
        {

            [Test]
            public void Peaks_Common_Array_Object()
            {

                try
                {

                    var coll = new String[] {"a", "b", "c"};

                    var result = Arrays.Peaks(coll);

                    Assert.IsTrue(result.Count() == 1);
                    Assert.IsTrue(result[0] == "c");

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }

            }

            [Test]
            public void Peaks_Nullable_Null_ArraysException()
            {

                try
                {

                    String[] coll = null;

                    Assert.Throws(typeof(ArraysException), delegate { Arrays.Peaks(coll); });

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }

            }

            [Test]
            public void Peaks_Special_Array_ArraysException()
            {

                try
                {

                    var coll = new String[] {null, null, null};

                    Assert.Throws(typeof(ArraysException), delegate { Arrays.Peaks(coll); });

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }

            }

            [Test]
            public void Peaks_Special_Array_ArraysException__1()
            {

                try
                {

                    var coll = new String[0];

                    Assert.Throws(typeof(ArraysException), delegate { Arrays.Peaks(coll); });

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }

            }

            [Test]
            public void Peaks_Special_Array_Object__2()
            {

                try
                {

                    var coll = new String[] {"a", "b", null};

                    var result = Arrays.Peaks(coll);

                    Assert.IsTrue(result.Count() == 1);
                    Assert.IsTrue(result[0].Equals("b"));

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }

            }

        }

        class Troughs
        {

            [Test]
            public void Troughs_Common_Array_Object()
            {

                try
                {

                    var coll = new String[] {"a", "b", "c"};

                    var result = Arrays.Troughs(coll);

                    Assert.IsTrue(result.Count() == 1);
                    Assert.IsTrue(result[0] == "a");

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }

            }

            [Test]
            public void Troughs_Nullable_Null_ArraysException()
            {

                try
                {

                    String[] coll = null;

                    Assert.Throws(typeof(ArraysException), delegate { Arrays.Troughs(coll); });

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }

            }

            [Test]
            public void Troughs_Special_Array_ArraysException()
            {

                try
                {

                    var coll = new String[] {null, null, null};

                    Assert.Throws(typeof(ArraysException), delegate { Arrays.Troughs(coll); });

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }

            }

            [Test]
            public void Troughs_Special_Array_ArraysException__1()
            {

                try
                {

                    var coll = new String[0];

                    Assert.Throws(typeof(ArraysException), delegate { Arrays.Troughs(coll); });

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }

            }

            [Test]
            public void Troughs_Special_Array_Object__2()
            {

                try
                {

                    var coll = new String[] {"a", "b", null};

                    var result = Arrays.Troughs(coll);

                    Assert.IsTrue(result.Count() == 1);
                    Assert.IsTrue(result[0].Equals("a"));

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }

            }

        }
    }
    
}

File: Java/test/tech/getrealgood/util/test/ArraysTests.java
package tech.getrealgood.util.test;

import org.junit.jupiter.api.*;
import org.junit.platform.commons.logging.Logger;
import org.junit.platform.commons.logging.LoggerFactory;
import tech.getrealgood.util.ArraysException;

import static org.junit.jupiter.api.Assertions.*;

class ArraysTests {

    private static final Logger log = LoggerFactory.getLogger(ArraysTests.class);

    @Nested
    class IsEmptyTests {

        @Test
        public void isEmpty_Common_Array_Boolean() {

            try {
                var o = new String[]{"sample1", "sample2"};
                assertEquals(false, tech.getrealgood.util.Arrays.isEmpty(o));

            } catch (Exception ex) {
                fail(ex.getMessage());
            }
        }

        @Test
        public void isEmpty_Common_Array_Boolean__1() {

            try {
                var o = new String[0];
                assertEquals(true, tech.getrealgood.util.Arrays.isEmpty(o));

            } catch (Exception ex) {
                fail(ex.getMessage());
            }
        }

        @Test
        public void isEmpty_Special_Array_Boolean() {

            try {
                var o = new String[]{"", ""};
                assertEquals(false, tech.getrealgood.util.Arrays.isEmpty(o));

            } catch (Exception ex) {
                fail(ex.getMessage());
            }
        }

        @Test
        public void isEmpty_Special_Array_Boolean__1() {

            try {
                var o = new String[]{null, null};
                assertEquals(false, tech.getrealgood.util.Arrays.isEmpty(o));

            } catch (Exception ex) {
                fail(ex.getMessage());
            }
        }

        @Test
        public void isEmpty_Nullable_Null_ArraysException() {

            try {
                String[] o = null;
                assertThrows(ArraysException.class, () -> {
                    tech.getrealgood.util.Arrays.isEmpty(o);
                });

            } catch (Exception ex) {
                fail(ex.getMessage());
            }
        }

    }

    @Nested
    class IsNotEmptyTests {

        @Test
        public void isNotEmpty_Common_Array_Boolean() {

            try {
                var o = new String[]{"sample1", "sample2"};
                assertEquals(true, tech.getrealgood.util.Arrays.isNotEmpty(o));

            } catch (Exception ex) {
                fail(ex.getMessage());
            }
        }

        @Test
        public void isNotEmpty_Common_Array_Boolean__1() {

            try {
                var o = new String[0];
                assertEquals(false, tech.getrealgood.util.Arrays.isNotEmpty(o));

            } catch (Exception ex) {
                fail(ex.getMessage());
            }
        }

        @Test
        public void isNotEmpty_Special_Array_Boolean() {

            try {
                var o = new String[]{"", ""};
                assertEquals(true, tech.getrealgood.util.Arrays.isNotEmpty(o));

            } catch (Exception ex) {
                fail(ex.getMessage());
            }
        }

        @Test
        public void isNotEmpty_Special_Array_Boolean__1() {

            try {
                var o = new String[]{null, null};
                assertEquals(true, tech.getrealgood.util.Arrays.isNotEmpty(o));

            } catch (Exception ex) {
                fail(ex.getMessage());
            }
        }

        @Test
        public void isNotEmpty_Nullable_Null_ArraysException() {

            try {
                String[] o = null;
                assertThrows(ArraysException.class, () -> {
                    tech.getrealgood.util.Arrays.isNotEmpty(o);
                });

            } catch (Exception ex) {
                fail(ex.getMessage());
            }
        }

    }


    @Nested
    class HasNullsAtTests {

        @Test
        public void hasNullsAt_Common_Array_List() {
            try {

                String[] s = new String[]{null, null, "null"};

                var result = tech.getrealgood.util.Arrays.hasNullsAt(s);
                assertEquals(2, result.size());

            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

        @Test
        public void hasNullsAt_Common_Collection_List__1() {
            try {

                String[] s = new String[]{"", "null", "null"};

                var result = tech.getrealgood.util.Arrays.hasNullsAt(s);
                assertEquals(0, result.size());

            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

        @Test
        public void hasNullsAt_Nullable_Null_ArraysException() {
            try {

                String[] s = null;

                assertThrows(ArraysException.class, () -> {
                    tech.getrealgood.util.Arrays.hasNullsAt(s);
                });

            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

        @Test
        public void hasNullsAt_Special_Array_List() {
            try {
                String[] s = new String[]{null, null, null};

                var result = tech.getrealgood.util.Arrays.hasNullsAt(s);
                assertEquals(3, result.size());

            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

    }

    @Nested
    class HasNullsTests {

        @Test
        public void hasNulls_Common_Array_Boolean() {
            try {

                String[] s = new String[]{"null", null, null};

                var result = tech.getrealgood.util.Arrays.hasNulls(s);
                assertEquals(true, result);

            } catch (Exception ex) {
                fail(ex.getMessage());
            }
        }

        @Test
        public void hasNulls_Common_Array_Boolean__1() {
            try {

                String[] s = new String[]{"null", "null", "null"};

                var result = tech.getrealgood.util.Arrays.hasNulls(s);
                assertEquals(false, result);

            } catch (Exception ex) {
                fail(ex.getMessage());
            }
        }

        @Test
        public void hasNulls_Special_Array_Boolean() {
            try {
                String[] s = new String[]{null, null, null};

                var result = tech.getrealgood.util.Arrays.hasNulls(s);
                assertEquals(true, result);

            } catch (Exception ex) {
                fail(ex.getMessage());
            }
        }

        @Test
        public void hasNulls_Nullable_Null_ArraysException() {
            try {

                String[] s = null;

                assertThrows(ArraysException.class, () -> {
                    tech.getrealgood.util.Arrays.hasNulls(s);
                });

            } catch (Exception ex) {
                fail(ex.getMessage());
            }
        }

    }

    @Nested
    class HasNullsCountTests {

        @Test
        public void hasNullsCount_Common_Array_Long() {
            try {

                String[] s = new String[]{"null", null, null};

                var result = tech.getrealgood.util.Arrays.hasNullsCount(s);
                assertEquals(2, result);

            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

        @Test
        public void hasNullsCount_Common_Array_Long__1() {
            try {

                String[] s = new String[]{"null", "null", "null"};

                var result = tech.getrealgood.util.Arrays.hasNullsCount(s);
                assertEquals(0, result);

            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

        @Test
        public void hasNullsCount_Nullable_Null_ArraysException() {
            try {

                String[] s = null;

                assertThrows(ArraysException.class, () -> {
                    tech.getrealgood.util.Arrays.hasNullsCount(s);
                });

            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

        @Test
        public void hasNullsCount_Special_Array_Long() {
            try {

                String[] s = new String[]{null, null, null};

                var result = tech.getrealgood.util.Arrays.hasNullsCount(s);
                assertEquals(3, result);

            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }


    }

    @Nested
    class MaxTests {

        @Test
        void max_Common_Collection_Object() {

            try {

                var coll = new String[]{"a", "b", "c"};

                var result = tech.getrealgood.util.Arrays.max(coll);

                assertTrue(result.equals("c"));

            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }


        @Test
        void max_Nullable_Null_ArraysException() {

            try {

                String[] coll = null;

                assertThrows(ArraysException.class, () -> {
                    tech.getrealgood.util.Arrays.max(coll);
                });

            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

        @Test
        void max_Special_Arrays_ArraysException() {

            try {

                var coll = new String[]{null, null, null};

                assertThrows(ArraysException.class, () -> {
                    tech.getrealgood.util.Arrays.max(coll);
                });

            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

        @Test
        void max_Special_Array_ArraysException__1() {

            try {

                var coll = new String[0];

                assertThrows(ArraysException.class, () -> {
                    tech.getrealgood.util.Arrays.max(coll);
                });

            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

        @Test
        void max_Special_Array_Object__2() {

            try {

                var coll = new String[]{"a", "b", null};

                var result = tech.getrealgood.util.Arrays.max(coll);

                assertTrue(result.equals("b"));

            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }


    }


    @Nested
    class MinTests {

        @Test
        void min_Common_Array_Object() {

            try {

                var coll = new String[]{"a", "b", "c"};

                var result = tech.getrealgood.util.Arrays.min(coll);

                assertTrue(result.equals("a"));

            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }


        @Test
        void min_Nullable_Null_ArraysException() {

            try {

                String[] coll = null;

                assertThrows(ArraysException.class, () -> {
                    tech.getrealgood.util.Arrays.min(coll);
                });

            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

        @Test
        void min_Special_Array_ArraysException() {

            try {

                var coll = new String[]{null, null, null};

                assertThrows(ArraysException.class, () -> {
                    tech.getrealgood.util.Arrays.min(coll);
                });

            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

        @Test
        void min_Special_Array_ArraysException__1() {

            try {

                var coll = new String[0];

                assertThrows(ArraysException.class, () -> {
                    tech.getrealgood.util.Arrays.min(coll);
                });

            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

        @Test
        void min_Special_Array_Object__2() {

            try {

                var coll = new String[]{"a", "b", null};

                var result = tech.getrealgood.util.Arrays.min(coll);

                assertTrue(result.equals("a"));

            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

    }

    @Nested
    class PeaksTests {

        @Test
        void peaks_Common_Array_Object() {

            try {

                var coll = new String[]{"a", "b", "c"};

                var result = tech.getrealgood.util.Arrays.peaks(coll);

                assertTrue(result.size() == 1);
                assertTrue(result.get(0) == "c");

            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

        @Test
        void peaks_Nullable_Null_ArraysException() {

            try {

                String[] coll = null;

                assertThrows(ArraysException.class, () -> {
                    tech.getrealgood.util.Arrays.peaks(coll);
                });

            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

        @Test
        void peaks_Special_Array_ArraysException() {

            try {

                var coll = new String[]{null, null, null};

                assertThrows(ArraysException.class, () -> {
                    tech.getrealgood.util.Arrays.peaks(coll);
                });

            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

        @Test
        void peaks_Special_Array_ArraysException__1() {

            try {

                var coll = new String[0];

                assertThrows(ArraysException.class, () -> {
                    tech.getrealgood.util.Arrays.peaks(coll);
                });

            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

        @Test
        void peaks_Special_Array_Object__2() {

            try {

                var coll = new String[]{"a", "b", null};

                var result = tech.getrealgood.util.Arrays.peaks(coll);

                assertTrue(result.size() == 1);
                assertTrue(result.get(0).equals("b"));

            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

    }


    @Nested
    class TroughsTests {

        @Test
        void troughs_Common_Array_Object() {

            try {

                var coll = new String[]{"a", "b", "c"};

                var result = tech.getrealgood.util.Arrays.troughs(coll);

                assertTrue(result.size() == 1);
                assertTrue(result.get(0) == "a");

            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }


        @Test
        void troughs_Nullable_Null_ArraysException() {

            try {

                String[] coll = null;

                assertThrows(ArraysException.class, () -> {
                    tech.getrealgood.util.Arrays.troughs(coll);
                });

            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

        @Test
        void troughs_Special_Array_ArraysException() {

            try {

                var coll = new String[]{null, null, null};

                assertThrows(ArraysException.class, () -> {
                    tech.getrealgood.util.Arrays.troughs(coll);
                });

            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

        @Test
        void troughs_Special_Array_ArraysException__1() {

            try {

                var coll = new String[0];

                assertThrows(ArraysException.class, () -> {
                    tech.getrealgood.util.Arrays.troughs(coll);
                });

            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

        @Test
        void troughs_Special_Array_Object__2() {

            try {

                var coll = new String[]{"a", "b", null};

                var result = tech.getrealgood.util.Arrays.troughs(coll);

                assertTrue(result.size() == 1);
                assertTrue(result.get(0).equals("a"));

            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

    }

}